source/common/network/listener_filter_buffer_impl.cc

```diff
@@ -0,0 +1,107 @@
+#include "source/common/network/listener_filter_buffer_impl.h"
+
+#include <string>
+
+namespace Envoy {
+namespace Network {
+
+ListenerFilterBufferImpl::ListenerFilterBufferImpl(IoHandle& io_handle,
+                                                   Event::Dispatcher& dispatcher,
+                                                   ListenerFilterBufferOnCloseCb close_cb,
+                                                   ListenerFilterBufferOnDataCb on_data_cb,
+                                                   uint64_t buffer_size)
+    : io_handle_(io_handle), dispatcher_(dispatcher), on_close_cb_(close_cb),
+      on_data_cb_(on_data_cb), buffer_(std::make_unique<char[]>(buffer_size)), base_(buffer_.get()),
+      buffer_size_(buffer_size) {
+  // If the buffer_size not greater than 0, it means that doesn't expect any data.
+  ASSERT(buffer_size > 0);
+
+  io_handle_.initializeFileEvent(
+      dispatcher_, [this](uint32_t events) { onFileEvent(events); },
+      Event::PlatformDefaultTriggerType, Event::FileReadyType::Read | Event::FileReadyType::Closed);
```

Comment:
@davinci26 I know why the Windows failed, it is due to I registered Read and Closed event at the same time.

After the initial data peek at 
https://github.com/envoyproxy/envoy/pull/17395/files#diff-6708370e112a5097f06e16009c3730b34f5ab0ee4f1e736f8816848409435888R98

The recv return try again.
But the Read event will be removed by this line https://github.com/envoyproxy/envoy/blob/46ab91335aa97866f9940eb10d0772b5c29e0db3/source/common/event/file_event_impl.cc#L140

So the test timeout in the end.

Do you know why we can't register Read and Closed events at the same time?

I saw the TlsInspect did register the Read and Closed events at the same time, I don't know how it works in the past.
https://github.com/envoyproxy/envoy/pull/17395/files#diff-3004fa07320c547f4fdd1ba336a346a906d1c4d8b3260216d6625f86dfcf85b5L117

Reply:
We don't register both read and close events because connection_impl does not handle this case. They are not handled because the remote close event`EPOLLRDHUP`( see [man](https://man7.org/linux/man-pages/man2/epoll_ctl.2.html)) triggers a read event anyway. So the connection implementation handles the case where read returns 0.

The behavior of tls inspector did not affect the behavior of the main datapath since it peeked the message it did not affect the event registration that the connection implementation would see.

Reply:
@davinci26 thanks for your replay. I will update the PR and remove the closed event.

For tls inspect, I think it still have a chance to face the same issue here. Althought first peek is fine, since it is before the event registration. But if the second peek (after the event registration) doesn't return enough data, so I should be able stuck at waiting for read event forever.

I submitted another PR to add assert on both read and closed event registered to avoid people get into this again.
https://github.com/envoyproxy/envoy/pull/18265

Also I saw the ConnectionImpl has the handle of close by read call. Is there something more missing in the ConnectionImpl?
https://github.com/envoyproxy/envoy/blob/f041df430c04d7bfd8d1954e506fe150a2443a05/source/common/network/connection_impl.cc#L624-L627

https://github.com/envoyproxy/envoy/blob/f041df430c04d7bfd8d1954e506fe150a2443a05/source/common/network/raw_buffer_socket.cc#L25-L28